src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs

```diff
@@ -886,31 +888,43 @@ public static ReadOnlyCollection<TimeZoneInfo> GetSystemTimeZones()
 
             lock (cachedData)
             {
+                if (cachedData._systemTimeZonesSorted)
+                {
+                    Debug.Assert(cachedData._readOnlySystemTimeZones != null);
+                    return cachedData._readOnlySystemTimeZones;
+                }
+
                 if (cachedData._readOnlySystemTimeZones == null)
                 {
                     PopulateAllSystemTimeZones(cachedData);
                     cachedData._allSystemTimeZonesRead = true;
+                }
 
-                    if (cachedData._systemTimeZones != null)
-                    {
-                        // return a collection of the cached system time zones
-                        TimeZoneInfo[] array = new TimeZoneInfo[cachedData._systemTimeZones.Count];
-                        cachedData._systemTimeZones.Values.CopyTo(array, 0);
-
-                        // sort and copy the TimeZoneInfo's into a ReadOnlyCollection for the user
-                        Array.Sort(array, static (x, y) =>
-                        {
-                            // sort by BaseUtcOffset first and by DisplayName second - this is similar to the Windows Date/Time control panel
-                            int comparison = x.BaseUtcOffset.CompareTo(y.BaseUtcOffset);
-                            return comparison == 0 ? string.CompareOrdinal(x.DisplayName, y.DisplayName) : comparison;
-                        });
+                if (cachedData._systemTimeZones == null)
+                {
+                    cachedData._readOnlySystemTimeZones = ReadOnlyCollection<TimeZoneInfo>.Empty;
+                }
+                else if (Unsorted)
+                {
+                    List<TimeZoneInfo> unsortedSystemTimeZones = new List<TimeZoneInfo>(cachedData._systemTimeZones.Values);
+                    cachedData._readOnlySystemTimeZones = new ReadOnlyCollection<TimeZoneInfo>(unsortedSystemTimeZones);
+                }
+                else
```

Comment:
why all this change? isn't enough only change line 924 to test `Unsorted` and decide to use the Id or DisplayName. 

I am not with not sorting the list at all using the offset. I think the perf hit coming from display name which is we try to avoid here.

Reply:
I was trying to cover the scenario in which a user grabs an unsorted collection and then changes their mind and grabs a sorted collection.

Reply:
With that in mind, for the second call, the scenarios in which `cachedData._readOnlySystemTimeZones != null` are
1. Empty case
2. Unsorted Case
3. Sorted Case

And the topmost early return I was thinking if users already did the work to calculate the sorted collection, only return the sorted collection if there are future calls, ignoring calls to get an unsorted collection (I dont know what value there would be in allowing users to get the unsorted collection after already having a sorted collection as at that point they already hit the performance cost).

Reply:
Sorry, I forgot to mention that a note in the description was that I figured since part of the point of this version was to be the most performant version of `GetSystemTimeZones` that it would just avoid sorting altogether, so the users wont incur any extra performance cost. It turned out though, that the unsorted list is mostly sorted by IDs except for having `UTC` at the beginning.